File: sdk/src/main/java/org/coralibre/android/sdk/internal/datatypes/util/DiagnosisKeyUtil.java
package org.coralibre.android.sdk.internal.datatypes.util;

import org.coralibre.android.sdk.internal.database.persistent.entity.EntityDiagnosisKey;
import org.coralibre.android.sdk.internal.datatypes.DiagnosisKey;
import org.coralibre.android.sdk.internal.datatypes.ENInterval;
import org.coralibre.android.sdk.internal.datatypes.TemporaryExposureKey_internal;
import org.coralibre.android.sdk.proto.TemporaryExposureKeyFile;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class DiagnosisKeyUtil {


    public static List<DiagnosisKey> toDiagnosisKeys(List<TemporaryExposureKeyFile.TemporaryExposureKeyProto> tekProtos) {
        List<DiagnosisKey> entityDiagnosisKeys = new ArrayList<>();
        for (TemporaryExposureKeyFile.TemporaryExposureKeyProto tekProto : tekProtos) {
            if (!tekProto.hasKeyData()) {
                throw new IllegalArgumentException("missing tekProto keyData");
            } else if (!tekProto.hasRollingStartIntervalNumber()) {
                throw new IllegalArgumentException("missing tekProto rollingStartIntervalNumber");
            } else if (!tekProto.hasRollingPeriod()) {
                throw new IllegalArgumentException("missing tekProto rollingPeriod");
            }

            entityDiagnosisKeys.add(
                new DiagnosisKey(
                    new TemporaryExposureKey_internal(
                        new ENInterval(tekProto.getRollingStartIntervalNumber()),
                        tekProto.getKeyData().toByteArray()),
                    tekProto.hasTransmissionRiskLevel() ? tekProto.getTransmissionRiskLevel() : 0)
            );
        }
        return entityDiagnosisKeys;
    }

}

File: sdk/src/main/java/org/coralibre/android/sdk/internal/datatypes/util/DiagnosisKeyUtil.kt
package org.coralibre.android.sdk.internal.datatypes.util

import org.coralibre.android.sdk.internal.datatypes.DiagnosisKey
import org.coralibre.android.sdk.internal.datatypes.ENInterval
import org.coralibre.android.sdk.internal.datatypes.InternalTemporaryExposureKey
import org.coralibre.android.sdk.proto.TemporaryExposureKeyFile.TemporaryExposureKeyProto
import java.util.ArrayList

object DiagnosisKeyUtil {
    @JvmStatic
    fun toDiagnosisKeys(tekProtos: List<TemporaryExposureKeyProto>): List<DiagnosisKey> {
        val entityDiagnosisKeys: MutableList<DiagnosisKey> = ArrayList()
        for (tekProto in tekProtos) {
            require(tekProto.hasKeyData()) { "missing tekProto keyData" }
            require(tekProto.hasRollingStartIntervalNumber()) { "missing tekProto rollingStartIntervalNumber" }
            entityDiagnosisKeys.add(
                DiagnosisKey(
                    InternalTemporaryExposureKey(
                        ENInterval(tekProto.rollingStartIntervalNumber.toLong()),
                        tekProto.keyData.toByteArray()
                    ),
                    if (tekProto.hasTransmissionRiskLevel()) tekProto.transmissionRiskLevel else 0
                )
            )
        }
        return entityDiagnosisKeys
    }
}
